cmd/thv-operator/controllers/mcpserver_controller.go

```diff
@@ -219,6 +235,113 @@ func (r *MCPServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	return ctrl.Result{}, nil
 }
 
+// ensureRBACResource is a generic helper function to ensure a Kubernetes resource exists and is up to date
+func (r *MCPServerReconciler) ensureRBACResource(
+	ctx context.Context,
+	mcpServer *mcpv1alpha1.MCPServer,
+	resourceType string,
+	createResource func() client.Object,
+) error {
+	// Get current resource
+	current := createResource()
+	objectKey := types.NamespacedName{Name: mcpServer.Name, Namespace: mcpServer.Namespace}
+	err := r.Get(ctx, objectKey, current)
+
+	if err != nil {
```

Comment:
nit: you might make this function easier to read if you move the logic within this if/else to dedicated functions. it got quite long.